state-sync/src/coordinator.rs

```diff
@@ -1054,6 +1054,84 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
         })
     }
 
+    /// Calculates the next version and epoch to request (assuming the given transaction list
+    /// and ledger info will be applied successfully).  Note: if no ledger info is specified,
+    /// we assume the next chunk will be for our current epoch.
+    fn calculate_new_known_version_and_epoch(
+        &mut self,
+        txn_list_with_proof: TransactionListWithProof,
+        ledger_info: Option<LedgerInfoWithSignatures>,
+    ) -> Result<(u64, u64), Error> {
+        let new_version = self
+            .local_state
+            .synced_version()
+            .checked_add(txn_list_with_proof.len() as u64)
+            .ok_or_else(|| {
+                Error::IntegerOverflow("The new version to request has overflown!".into())
+            })?;
+
+        let mut new_epoch = self.local_state.trusted_epoch();
+        if let Some(ledger_info) = ledger_info {
+            if ledger_info.ledger_info().version() == new_version
+                && ledger_info.ledger_info().ends_epoch()
+            {
+                // This chunk is going to finish the current epoch. Choose the next one.
+                new_epoch = self
+                    .local_state
+                    .trusted_epoch()
+                    .checked_add(1)
+                    .ok_or_else(|| {
+                        Error::IntegerOverflow("The new epoch to request has overflown!".into())
+                    })?;
+            }
+        }
```

Comment:
Unfortunately rust doesn't yet support statements like `if let Some(x) = x && .... { .. }` on stable. Which means if I adopt this approach I'll have 3 branches instead of 2, and 2 of those branches will return the same thing, e.g.:
```
let epoch = if let ... {
  if ... {
    ...
  } else {
    self.local_state.trusted_epoch()
  }
} else {
  self.local_state.trusted_epoch()
}
```

Personally, I think how it is right now is clearer (2 branches as opposed to 3). Thoughts?